File: searchingAlgo/DepthFirstSearch/depthFirstSearch.kt
/**
 * Creates a Graph with [size] nodes
 * @param init constructor for adding the edges of the nodes delegated to the instances
 */
class Graph(size: Int, init: Graph.() -> Unit) {
    val adjacency: Array<MutableList<Int>> = Array(size) { ArrayList() }

    init { init() }

    fun addEdge(node: Int, to: Int) = adjacency[node].add(to)
}

/**
 * Prints the route taken when exploring the [graph] starting from [current]
 * @param searched an array of booleans relating to whether the relating nodes have been searched
 */
fun depthFirstSearch(graph: Graph, current: Int, searched: Array<Boolean> = Array(graph.adjacency.size) { false }) {
    if(!searched[current]) {
        searched[current] = true
        print("$current ")
        for (next in graph.adjacency[current]) {
            depthFirstSearch(graph, next, searched)
        }
    }
}

fun main() {
    val graph = Graph(4) {
        addEdge(0, 1)
        addEdge(0, 2)
        addEdge(1, 2)
        addEdge(2, 0)
        addEdge(2, 3)
        addEdge(3, 3)
    }
    println("Following is Depth First Traversal " + "(Starting from vertex 2)")
    depthFirstSearch(graph, 2)
}

File: searchingAlgo/depthFirstSearch/depthfirstsearch.cpp
#include<iostream>
#include <list>

using namespace std;


class Graph //create a Graph node
{
    int V;
    list<int> *adj;
    void DFS_visit(int v, bool visited[]);
public:
    Graph(int V);
    void addEdge(int v, int w);
    void DFS(int v);
};

Graph::Graph(int V) //constructor
{
    this->V = V;
    adj = new list<int>[V];
}

void Graph::addEdge(int v, int w)
{
    adj[v].push_back(w);
}

void Graph::DFS_visit(int v, bool visited[])
{

    visited[v] = true;
    cout << v << " ";
    list<int>::iterator i;
    for (i = adj[v].begin(); i != adj[v].end();i++)
        {
            if (!visited[*i])
            DFS_visit(*i, visited);
        }
}


void Graph::DFS(int v)
{
    // Marking all the vertices as not visited
    bool *visited = new bool[V];
    for (int i = 0; i < V; i++)
        visited[i] = false;


  DFS_visit(v, visited);
}

int main()
{
    Graph g(4);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 0);
    g.addEdge(2, 3);
    g.addEdge(3, 3);

    cout << "Following is Depth First Traversal \n";
    int n;
    cin>>n;
    g.DFS(n);

    return 0;
}
//INPUT: 5
//OUTPUT:
//Following is Depth First Traversal 
//5 2 0 1 3
